composer/callbacks/run_directory_uploader.py

```diff
@@ -0,0 +1,323 @@
+        self._finished = self._finished_cls()
+        self._last_upload_timestamp = 0.0
+        self._workers = [
+            self._proc_class(target=_upload_worker,
+                             kwargs={
+                                 "file_queue": self._file_upload_queue,
+                                 "is_finished": self._finished,
+                                 "upload_staging_dir": self._upload_staging_folder,
+                                 "provider_name": self._provider,
+                                 "container_name": self._container,
+                                 "object_name_prefix": self._object_name_prefix,
+                                 "init_kwargs": self._provider_init_kwargs,
+                             }) for _ in range(self._num_concurrent_uploads)
+        ]
+        for worker in self._workers:
+            worker.start()
+
+    def batch_end(self, state: State, logger: Logger) -> None:
+        if get_run_directory() is None:
+            return
+        if (state.batch_idx + 1) % self._upload_every_n_batches == 0:
+            self._trigger_upload(logger, LogLevel.BATCH)
+
+    def epoch_end(self, state: State, logger: Logger) -> None:
+        del state  # unused
+        if get_run_directory() is None:
+            return
+        self._trigger_upload(logger, LogLevel.EPOCH)
+
+    def training_end(self, state: State, logger: Logger) -> None:
+        del state  # unused
+        if get_run_directory() is None:
+            return
+        self._trigger_upload(logger, LogLevel.FIT)
+
+    def post_close(self):
+        # Cleaning up on post_close to ensure that all artifacts are uploaded
+        self._trigger_upload(logger=None, log_level=None)
+        if self._finished is not None:
+            self._finished.set()
+        for worker in self._workers:
+            worker.join()
+
+    def _trigger_upload(self, logger: Optional[Logger], log_level: Optional[LogLevel]) -> None:
+        # Ensure that every rank is at this point
+        # Assuming only the main thread on each rank writes to the run directory, then the barrier here will ensure
+        # that the run directory is not being modified after we pass this barrier
+        ddp.barrier()
+        new_last_uploaded_timestamp = time.time()
+        # Now, for each file that was modified since self._last_upload_timestamp, copy it to the temporary directory
+        run_directory = get_run_directory()
+        assert run_directory is not None, "invariant error"
+        files_to_be_uploaded = []
+
+        # check if any upload threads have crashed. if so, then shutdown the training process
+        for worker in self._workers:
+            if not worker.is_alive():
+                assert self._finished is not None, "invariant error"
+                self._finished.set()
+                raise RuntimeError("Upload worker crashed unexpectedly")
+        for root, dirs, files in os.walk(run_directory):
+            del dirs  # unused
```

Comment:
```suggestion
        for root, _, files in os.walk(run_directory):
```

Reply:
Google style guide prefers `del` instead of `_`: https://google.github.io/styleguide/pyguide.html#214-decision